Stripe SDK retry on java.net.ConnectException

Stripe's Java SDK has had a retry mechanism for a little while. Looking at the source, it'll retry on exceptions that are instances of `java.net.ConnectException`.
```
    // Retry on connection error.
    if ((exception != null)
        && (exception.getCause() != null)
        && (exception.getCause() instanceof ConnectException)) {
      return true;
    }
```
However, we occasionally see in our logs the following stack trace:
```
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_201]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_201]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_201]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_201]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_201]
	at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_201]
	at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:666) ~[na:1.8.0_201]
	at sun.net.NetworkClient.doConnect(NetworkClient.java:175) ~[na:1.8.0_201]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:463) ~[na:1.8.0_201]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:558) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:367) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492) ~[na:1.8.0_201]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347) ~[na:1.8.0_201]
	at com.stripe.net.HttpURLConnectionClient.request(HttpURLConnectionClient.java:37) ~[stripe-java-19.24.0.jar:19.24.0]
	... 84 common frames omitted
```
The exception here is a `java.net.SocketTimeoutException` so the retry logic is not triggered and the call fails.

Any idea what could be causing those `java.net.SocketTimeoutException`s? Should the HTTP client retry when encountering them?

--- a/src/main/java/com/stripe/net/HttpClient.java
+++ b/src/main/java/com/stripe/net/HttpClient.java
@@ -5,6 +5,7 @@ import com.stripe.exception.ApiConnectionException;
 import com.stripe.exception.StripeException;
 import com.stripe.util.Stopwatch;
 import java.net.ConnectException;
+import java.net.SocketTimeoutException;
 import java.time.Duration;
 import java.util.HashMap;
 import java.util.Map;
@@ -174,7 +175,7 @@ public abstract class HttpClient {
     // Retry on connection error.
     if ((exception != null)
         && (exception.getCause() != null)
-        && (exception.getCause() instanceof ConnectException)) {
+        && (exception.getCause() instanceof ConnectException || exception.getCause() instanceof SocketTimeoutException)) {
       return true;
     }
 

--- a/src/main/java/com/stripe/net/HttpClient.java
+++ b/src/main/java/com/stripe/net/HttpClient.java
@@ -175,7 +175,8 @@ public abstract class HttpClient {
     // Retry on connection error.
     if ((exception != null)
         && (exception.getCause() != null)
-        && (exception.getCause() instanceof ConnectException || exception.getCause() instanceof SocketTimeoutException)) {
+        && (exception.getCause() instanceof ConnectException
+            || exception.getCause() instanceof SocketTimeoutException)) {
       return true;
     }
 

--- a/src/test/java/com/stripe/net/HttpClientTest.java
+++ b/src/test/java/com/stripe/net/HttpClientTest.java
@@ -77,7 +77,8 @@ public class HttpClientTest extends BaseStripeTest {
   @Test
   public void testRequestWithRetriesSocketTimeoutException() throws StripeException {
     Mockito.when(this.client.request(this.request))
-        .thenThrow(new ApiConnectionException("foo", new SocketTimeoutException("timeout or something")))
+        .thenThrow(
+            new ApiConnectionException("foo", new SocketTimeoutException("timeout or something")))
         .thenReturn(new StripeResponse(200, emptyHeaders, "{}"));
 
     StripeResponse response = this.client.requestWithRetries(this.request);

--- a/src/test/java/com/stripe/net/HttpClientTest.java
+++ b/src/test/java/com/stripe/net/HttpClientTest.java
@@ -12,6 +12,7 @@ import com.stripe.BaseStripeTest;
 import com.stripe.exception.ApiConnectionException;
 import com.stripe.exception.StripeException;
 import java.net.ConnectException;
+import java.net.SocketTimeoutException;
 import java.util.Collections;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -70,6 +71,39 @@ public class HttpClientTest extends BaseStripeTest {
     assertEquals("3", e.getMessage());
     assertNotNull(e.getCause());
     assertTrue(e.getCause() instanceof ConnectException);
+    assertEquals("timeout 3", e.getCause().getMessage());
+  }
+
+  @Test
+  public void testRequestWithRetriesSocketTimeoutException() throws StripeException {
+    Mockito.when(this.client.request(this.request))
+        .thenThrow(new ApiConnectionException("foo", new SocketTimeoutException("timeout or something")))
+        .thenReturn(new StripeResponse(200, emptyHeaders, "{}"));
+
+    StripeResponse response = this.client.requestWithRetries(this.request);
+
+    assertNotNull(response);
+    assertEquals(200, response.code());
+    assertEquals(1, response.numRetries());
+  }
+
+  @Test
+  public void testRequestWithRetriesSocketTimeoutExceptionRethrowAfterAllAttempts()
+      throws StripeException {
+    Mockito.when(this.client.request(this.request))
+        .thenThrow(new ApiConnectionException("1", new SocketTimeoutException("timeout 1")))
+        .thenThrow(new ApiConnectionException("2", new SocketTimeoutException("timeout 2")))
+        .thenThrow(new ApiConnectionException("3", new SocketTimeoutException("timeout 3")));
+
+    ApiConnectionException e =
+        assertThrows(
+            ApiConnectionException.class,
+            () -> {
+              this.client.requestWithRetries(this.request);
+            });
+    assertEquals("3", e.getMessage());
+    assertNotNull(e.getCause());
+    assertTrue(e.getCause() instanceof SocketTimeoutException);
     assertEquals("timeout 3", e.getCause().getMessage());
   }
 

